ReadData and DataBlockCodec

tpietzsch opened the pull request:
Add `DataBlock` methods
`void readData(ByteOrder, ReadData)` and
`void writeData(ByteOrder, OutputStream)`
remove
`ByteBuffer toByteBuffer()` and
`void readData(ByteBuffer)`.

Add `Splittable.ReadData`, etc.

Still not fully settled...

cmhulbert commented:
I think there is some overlap with this branch, I'll be back at work on Monday fully, so hopefully we can distill the important parts from these two. My branch is more focused on getting this working for sharding/codecs, but hopefully there is a lot of shared logic. I created a draft PR to better compare at #138

Some specific files that are likely to overlap:
[SplitableData.java](https://github.com/saalfeldlab/n5/blob/8042d4379622bc6a97806e7bf3bfdb0abaaa2f27/src/main/java/org/janelia/saalfeldlab/n5/SplitableData.java)
[SplitKeyValueAccessData.java](https://github.com/saalfeldlab/n5/blob/8042d4379622bc6a97806e7bf3bfdb0abaaa2f27/src/main/java/org/janelia/saalfeldlab/n5/SplitKeyValueAccessData.java)
[SplitByteBufferedData.java](https://github.com/saalfeldlab/n5/blob/8042d4379622bc6a97806e7bf3bfdb0abaaa2f27/src/main/java/org/janelia/saalfeldlab/n5/SplitByteBufferedData.java)

tpietzsch commented:
I'm still working on this, so please wait for a bit (just until later today, probably) before starting to reconcile.

In this PR, the `ReadData` is turning out more of an abstraction over `byte[]`, `ByteBuffer`, `InputStream`, `OutputStream`. I'm changing the API to just always hand around `ReadData`. I'll write more about it when it has settled.
So that there is also `SplittableReadData` (extending `ReadData`) is not the important thing. It's more about cleaning up the `DataBlock/Compression/BlockWriter/BlockReader` API.

Probably we should set up a zoom meeting to discuss, later this week

cmhulbert commented:
Sounds good to me, thanks for taking a stab at this. I'll clean up some of what I have, and let me know when you have time to chat, would be happy to zoom

tpietzsch commented:
related PRs:

https://github.com/JaneliaSciComp/n5-zstandard/pull/4
https://github.com/saalfeldlab/n5-blosc/pull/11
https://github.com/saalfeldlab/n5-zarr/pull/63

tpietzsch commented:
Ok, I'm happy now.

The `ReadData` abstraction is used throughout, now:

[`DataBlock` de/serialization](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/DataBlock.java#L72-L102):
```java
public interface DataBlock<T> {
    void readData(ReadData readData) throws IOException;
    ReadData writeData(ByteOrder byteOrder);
}
```

[`BytesCodec`/`Compression` interface](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/BytesCodec.java#L6-L41):
```java
public interface BytesCodec {
    ReadData decode(ReadData readData, int decodedLength) throws IOException;
    ReadData encode(ReadData readData) throws IOException;
}
```

[block reader](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/DefaultBlockReader.java#L81-L86):
```java
    final ReadData data = ReadData.from(in)
                .decode(datasetAttributes.getCompression(), numBytes);
    dataBlock.readData(data);
```
and [block writer](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/DefaultBlockWriter.java#L81-L83):
```java
    dataBlock.writeData(ByteOrder.BIG_ENDIAN)
                .encode(datasetAttributes.getCompression())
                .writeTo(out);
```

There is also extensive javadoc for [ReadData](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/readdata/ReadData.java), [BytesCodec](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/BytesCodec.java).

@axtimwalde @bogovicj @cmhulbert Let's set up a review session this week? I think this is ready to merge.

bogovicj reviewed (commented):
A few small comments / observations

tpietzsch commented on src/main/java/org/janelia/saalfeldlab/n5/CompressionAdapter.java:
```diff
@@ -78,6 +78,9 @@ public static synchronized void update(final boolean override) {
 			final ClassLoader classLoader = Thread.currentThread().getContextClassLoader();
 			final Index<CompressionType> annotationIndex = Index.load(CompressionType.class, classLoader);
+			for (final IndexItem<CompressionType> item : annotationIndex) {
+				System.out.println("item.className() = " + item.className());
```
oops. no. that's left over from debugging....

tpietzsch commented on src/main/java/org/janelia/saalfeldlab/n5/readdata/SplittableReadData.java:
```diff
@@ -0,0 +1,12 @@
+
+public interface SplittableReadData extends ReadData {
+
+	ReadData split(final long offset, final long length) throws IOException;
```
I don't know, because so far I didn't need splitting at all, but it sounds like that could be useful!

If the `ReadData` knows its length, then you can just `data.split(offset, data.length() - offset)`.

Split "to the end" would be helpful for `ReadData` that don't know their length (and to avoid having to retrieve it).
Maybe we just add `ReadData split(long offset)` without a `length` argument to split to the end.

At the moment, the javadoc of `ReadData.splittable()` says
> The returned SplittableReadData has a known length and multiple inputStreams can be opened on it.

That contract should be changed then, probably.
Maybe we should add an explicit `ReadData.materialize()` method for that instead.

bogovicj replied:
Something related I came across while experimenting.  How would you feel about a method `limit(long)`?
It should be equivalent to `split(0,long)`, when splittable, but could be useful in other cases, specifically:

```java
final InputStream ones = new InputStream() {
     @Override public int read() throws IOException { return 1; }
};

// This fails because `splittable` tries to read the whole (infinite) input stream which is reasonable.
ReadData firstTenBySpllit = ReadData.from(ones).splittable().split(0, 10);

// this could be a sensible alternative
ReadData firstTen = ReadData.from(ones).limit(10);
```

tpietzsch replied:
That is more difficult, I think.
Ideally `ReadData` should be considered immutable.
Conceptually, `ReadData.split(...)` shouldn't modify the `ReadData` that you call it on.

However
```java
final InputStream sequence = new InputStream() {
	private int value = 0;
	@Override public int read() throws IOException { return value++; }
};

ReadData data = ReadData.from(sequence);
// data.inputStream().read() would be 0 here

ReadData firstTen = data.limit(10);
// data.inputStream().read() would be 10 here
```

Therefore, a `limit()` implementation that reads a bit of the InputStream turns `data` unusable.

Maybe then, `limit(n)` should be considered destructive and return a pair of `ReadData`, one for the first `n` bytes, and one for the rest.

That would be an actual "split" then.
And maybe we should re-consider naming...
The current `split()` is actually more a "view" than a "split". Maybe `slice(...)` would be relatively consistent with `ByteBuffer` terminology?

tpietzsch replied:
... but I totally agree that `limit()` seems useful!

bogovicj replied:
Ah, I see what you mean, good point.

> Maybe then, limit(n) should be considered destructive and return a pair of ReadData, one for the first n bytes, and one for the rest.

> The current split() is actually more a "view" than a "split". Maybe slice(...) would be relatively consistent with ByteBuffer terminology?

This makes sense to me - I'd be happy with something like:

```java
final InputStream sequence = new InputStream() {
	private int value = 0;
	@Override public int read() throws IOException { return value++; }
};

ReadData data = ReadData.from(sequence);

SplitReadData split = data.split(10);  // basically a proposed "limit", a Pair<ReadData,ReadData>

ReadData firstTen = split.first();
ReadData lastInfinity = split.last();
ReadData view = firstTen.slice(5,5); // what is currently called "split"
```

tpietzsch commented:
I separated de/serialization out of `DataBlock<T>` completely, into `DataBlockCodec<T>`. (I had started to do the "incremental" version that we had discussed, but in the end just did the whole thing.)
The `DataBlock` implementations are [super](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/LongArrayDataBlock.java) [basic](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/StringDataBlock.java) now, e.g.
```java
public class LongArrayDataBlock extends AbstractDataBlock<long[]> {
	public LongArrayDataBlock(final int[] size, final long[] gridPosition, final long[] data) {
		super(size, gridPosition, data, a -> a.length);
	}
}

public class StringDataBlock extends AbstractDataBlock<String[]> {
	public StringDataBlock(final int[] size, final long[] gridPosition, final String[] data) {
		super(size, gridPosition, data, a -> a.length);
	}
}
```

The [`DataBlockCodec`](https://github.com/tpietzsch/n5/blob/readdata/src/main/java/org/janelia/saalfeldlab/n5/codec/DataBlockCodec.java) interface looks like this:
```java
public interface DataBlockCodec<T> {
	ReadData encode(DataBlock<T> dataBlock) throws IOException;
	DataBlock<T> decode(ReadData readData, long[] gridPosition) throws IOException;
}
```
That looks like it could be quite compatible with @bogovicj `DataBlockInputStream` ideas?
`DataBlockCodec` can be obtained from `DatasetAttributes`. (Which is very nice, because `ZarrDatasetAttributes` can override it, and consequently we can just use `DefaultBlockReader/Writer` for Zarr.)


As discussed in the last code review, I also did the following:
* I removed `ByteOrder` from `ReadData` again.
* I replaced the `OutputStreamEncoder` interface (bundling `OutputStream` and `finish` method) with a `ProxyOutputStream` wrapper that blocks the `close()` method. I couldn't use `UnaryOperator<OutputStream>` though, because I want to throw `IOException`.
* I removed the ReadData.encode/decode methods that just called `return Compression.encode/decode(this)`.
* I inlined the `BytesCodec` interface into `Compression` (for now).

cmhulbert commented:
Thanks @tpietzsch! I'll plan to spend some time today and tomorrow to try and rebase my current branch off this, and see how it plays with the codecs, especially when chaining arbitrary codecs. That should hopefully expose any thing else we need. Would also be happy to chat some time next week

bogovicj commented on src/main/java/org/janelia/saalfeldlab/n5/codec/N5Codecs.java:
```diff
@@ -0,0 +1,268 @@
+		}
+	}
+
+	static class ChunkHeader {
```
I'd call this `BlockHeader` since everything in N5 uses "block" not "chunk"

tpietzsch replied:
done

bogovicj commented on src/main/java/org/janelia/saalfeldlab/n5/readdata/ByteArraySplittableReadData.java:
```diff
@@ -0,0 +1,47 @@
+import java.io.InputStream;
+import java.util.Arrays;
+
+class ByteArraySplittableReadData implements ReadData {
```
maybe rename to `ByteArrayReadData` since we're not doing the splittable part yet?
I'm also happy to keep the name as is since the splitting will come soon.

bogovicj commented on src/main/java/org/janelia/saalfeldlab/n5/codec/N5Codecs.java:
```diff
@@ -0,0 +1,268 @@
+				final byte[] serializedData = flattenedArray.getBytes(ENCODING);
+				new ChunkHeader(dataBlock.getSize(), serializedData.length).writeTo(out);
+				compression.encode(ReadData.from(serializedData)).writeTo(out);
+				out.flush();
```
`out.flush()` is called in all the encode methods here - should it be the responsibility of the `OutputStreamOperator` to flush? could it instead be the job of `LazyReadData`? 

<strike>I'll make a note there too.</strike>

Edit: okay maybe I get it now.  `OutputStreamWriter` is a functional interface and can't be flushed itself.
This leads to a different question.

tpietzsch replied:
I revised where flush() happens. 👍

bogovicj commented on src/main/java/org/janelia/saalfeldlab/n5/readdata/LazyReadData.java:
```diff
@@ -0,0 +1,81 @@
+
+class LazyReadData implements ReadData {
+
+	LazyReadData(final OutputStreamWriter writer) {
```
I'm confused here. What's the benefit of using this functional interface instead of an `OutputStream` directly?

tpietzsch replied:
The `OutputStreamWriter` is a way to defer the writing/generation/serialization (not sure what to best call it) of the data until it is needed and it is clear what the destination is.

The `Compression` implementations (most of them) use `Input/OutputStream` wrappers to do the compression. The `DataBlockCodec::encode` methods use these `Compressions` so they have to use `OutputStream` too. Because they don't take an `OutputStream` argument we cannot serialize into an `OutputStream` directly.

Instead  `DataBlockCodec::encode` returns a `LazyReadData` that is not committed to any `OutputStream` yet --

- Maybe someone needs serialized data as a `byte[]` array. In that case, the `LazyReadData` will create a `ByteArrayOutputStream` and use the `OutputStreamWriter` to populate it.
- Maybe someone wants to write the serialized data into a `FileOutputStream`. In that case, the `LazyReadData` can pass that `FileOutputStream` to the `OutputStreamWriter` directly (without intermediate serialization into a `byte[]` array).

bogovicj reviewed (commented):
A few small ideas about names.

One bigger question about how `LazyReadData` works

bogovicj commented on src/main/java/org/janelia/saalfeldlab/n5/DataType.java:
```diff
@@ -44,10 +43,15 @@ public enum DataType {
-					gridPosition,
-					new byte[numElements])),
+			(blockSize, gridPosition, numElements) -> {
+				ByteArrayDataBlock dataBlock = new ByteArrayDataBlock(blockSize, gridPosition, new byte[numElements]);
```
this line looks like it can be deleted, and this block is stylistically different than other types.

tpietzsch replied:
saw it too late, but fixed it on master now
